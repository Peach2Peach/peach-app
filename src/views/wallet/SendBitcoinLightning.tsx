import { PaymentStatus } from "@breeztech/react-native-breez-sdk";
import bolt11 from "bolt11";
import { useState } from "react";
import { View } from "react-native";
import { Screen } from "../../components/Screen";
import { Button } from "../../components/buttons/Button";
import { BTCAmountInput } from "../../components/inputs/BTCAmountInput";
import { LightningInvoiceInput } from "../../components/inputs/LightningInvoiceInput";
import { useClosePopup, useSetPopup } from "../../components/popup/GlobalPopup";
import { PopupAction } from "../../components/popup/PopupAction";
import { PeachText } from "../../components/text/PeachText";
import { useValidatedState } from "../../hooks/useValidatedState";
import { ErrorPopup } from "../../popups/ErrorPopup";
import { SuccessPopup } from "../../popups/SuccessPopup";
import tw from "../../styles/tailwind";
import i18n from "../../utils/i18n";
import {
  MSAT_PER_SAT,
  useLightningWalletBalance,
} from "./hooks/useLightningWalletBalance";
import { usePayInvoice } from "./hooks/usePayInvoice";

const invoiceRules = { lightningInvoice: true };
export const SendBitcoinLightning = () => {
  const [invoice, setInvoice, isInvoiceValid, invoiceErrors] =
    useValidatedState<string>("", invoiceRules);
  const setPopup = useSetPopup();
  const closePopup = useClosePopup();
  const [amount, setAmount] = useState(0);
  const [paymentRequest, setPaymentRequest] =
    useState<ReturnType<typeof bolt11.decode>>();
  const { balance } = useLightningWalletBalance();
  const invoiceAmountMsat = paymentRequest?.millisatoshis
    ? Number(paymentRequest?.millisatoshis || 0)
    : amount * MSAT_PER_SAT;
  const hasEnoughBalance = balance.lightningMsats > invoiceAmountMsat;

  const updateInvoice = (value: string) => {
    setInvoice(value);
    try {
      const decodedInvoice = bolt11.decode(value);
      if (decodedInvoice?.millisatoshis) {
        setAmount(Number(decodedInvoice.millisatoshis) / MSAT_PER_SAT);
      } else {
        setAmount(0);
      }
      setPaymentRequest(decodedInvoice);
    } catch (e) {
      // do nothing
    }
  };
  const onFocus = () => setAmount(0);

  const { payInvoice, isPayingInvoice } = usePayInvoice({
    paymentRequest,
    amount: invoiceAmountMsat,
  });
  const payInvoiceAndDisplayStatus = async () => {
    const payment = await payInvoice();
    if (payment.status === PaymentStatus.FAILED) {
      return setPopup(
        <ErrorPopup
          title={i18n("wallet.sendBitcoin.lightningPaymentFailed.title")}
          content={
            <PeachText>
              {i18n("wallet.sendBitcoin.lightningPaymentFailed.text")}
              {"\n"}
              {payment.error}
            </PeachText>
          }
          actions={
            <PopupAction
              label={i18n("close")}
              iconId="xSquare"
              onPress={closePopup}
              reverseOrder
            />
          }
        />,
      );
    }
    return setPopup(
      <SuccessPopup
        title={i18n("wallet.sendBitcoin.lightningPaymentSuccess.title")}
        content={
          <PeachText>
            {i18n("wallet.sendBitcoin.lightningPaymentSuccess.text")}
          </PeachText>
        }
        actions={
          <PopupAction
            label={i18n("close")}
            iconId="xSquare"
            onPress={closePopup}
            reverseOrder
          />
        }
      />,
    );
  };
  return (
    <Screen header={i18n("wallet.sendBitcoin.title")}>
      <View style={[tw`gap-2 py-1`, tw`md:gap-8 md:py-6`]}>
        <LightningInvoiceInput
          value={invoice}
          accessibilityHint={i18n("form.address.lightningInvoice.label")}
          label={i18n("form.address.lightningInvoice.label")}
          required={false}
          onChangeText={updateInvoice}
          errorMessage={invoiceErrors}
        />
        {paymentRequest && (
          <BTCAmountInput
            accessibilityHint={i18n("form.lightningInvoice.amount.label")}
            value={String(amount)}
            onFocus={onFocus}
            editable={!paymentRequest.millisatoshis}
            onChangeText={(text) => setAmount(Number(text))}
            size="medium"
            textStyle={tw`absolute w-full py-0 opacity-0 grow h-38px input-text`}
            containerStyle={[
              tw`self-stretch justify-center px-2 py-3 overflow-hidden h-38px rounded-xl`,
              tw`border bg-primary-background-light border-black-65`,
            ]}
          />
        )}
        <Button
          onPress={payInvoiceAndDisplayStatus}
          loading={isPayingInvoice}
          disabled={
            !isInvoiceValid || invoiceAmountMsat === 0 || !hasEnoughBalance
          }
        >
          {i18n("wallet.sendBitcoin.payInvoice")}
        </Button>
      </View>
    </Screen>
  );
};
