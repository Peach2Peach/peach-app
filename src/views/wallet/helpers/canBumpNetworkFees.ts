import { PeachWallet } from "../../../utils/wallet/PeachWallet";

export const canBumpNetworkFees = (
  peachWallet: PeachWallet,
  transaction: TransactionSummary,
) =>
  !transaction.confirmed &&
  peachWallet.transactions
    .filter((tx) => !tx.confirmationTime?.height)
    .some(({ txid, sent }) => txid === transaction.id && sent > 0);
