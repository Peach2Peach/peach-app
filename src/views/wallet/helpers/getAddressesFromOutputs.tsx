import {
  Transaction as BitcoinTransaction,
  address as bitcoinAddress,
} from "bitcoinjs-lib";
import {
  Transaction as LiquidTransaction,
  address as liquidAddress,
} from "liquidjs-lib";
import { isDefined } from "../../../utils/validation/isDefined";
import { getLiquidNetwork } from "../../../utils/wallet/getLiquidNetwork";
import { getNetwork } from "../../../utils/wallet/getNetwork";

export const scriptToAddress = (script: Buffer, chain: string) => {
  try {
    return chain === "bitcoin"
      ? bitcoinAddress.fromOutputScript(script, getNetwork())
      : liquidAddress.fromOutputScript(script, getLiquidNetwork());
  } catch (e) {
    return undefined;
  }
};

export const getAddressesFromOutputs = ({
  outs,
  chain,
}: Pick<BitcoinTransaction | LiquidTransaction, "outs"> & { chain: Chain }) =>
  outs
    .map((v) => v.script)
    .filter((script) => script.byteLength > 1)
    .map((script) => scriptToAddress(script, chain))
    .filter(isDefined);
