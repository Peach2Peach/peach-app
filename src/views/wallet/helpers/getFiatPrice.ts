import { SATSINBTC } from "../../../constants";
import { round } from "../../../utils/math/round";

export const getFiatPrice = (amount: number, btcPrice: number) =>
  round(((btcPrice || 0) / SATSINBTC) * amount, 2).toFixed(2);
