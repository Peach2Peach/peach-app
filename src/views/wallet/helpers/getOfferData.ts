import { ContractSummary } from "../../../../peach-api/src/@types/contract";

export function getOfferData(
  offers: (SellOffer | BuyOffer)[],
  contracts: ContractSummary[],
  type: TransactionType,
) {
  return offers.map((offer) => {
    const contract = contracts.find((c) => c.id === offer.contractId);
    return {
      offerId: offer.id,
      contractId: offer.contractId,
      amount:
        contract?.amount ||
        (Array.isArray(offer.amount) ? offer.amount[0] : offer.amount),
      address:
        offer.type === "bid"
          ? offer.releaseAddress
          : type === "ESCROWFUNDED"
            ? offer.escrow || ""
            : offer.returnAddress,
      currency: contract?.currency,
      price: contract?.price,
    };
  });
}
