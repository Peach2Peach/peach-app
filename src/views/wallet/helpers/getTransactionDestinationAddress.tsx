import { getHighestValueOutputAddress } from '../../../utils/bitcoin'

const incomingTypes = ['REFUND', 'TRADE', 'DEPOSIT']
export const getTransactionDestinationAddress = (type: TransactionType, transactionDetails: Transaction) => {
  let address

  if (type === 'ESCROWFUNDED') {
    address = transactionDetails.vout.find(
      ({ scriptpubkey_type }) => scriptpubkey_type === 'v0_p2wsh',
    )?.scriptpubkey_address
  } else if (incomingTypes.includes(type)) {
    address = transactionDetails.vout.find(
      ({ scriptpubkey_type }) => scriptpubkey_type === 'v0_p2wpkh',
    )?.scriptpubkey_address
  }
  return address || getHighestValueOutputAddress(transactionDetails)
}
