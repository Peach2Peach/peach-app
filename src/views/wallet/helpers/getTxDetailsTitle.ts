import i18n from '../../../utils/i18n'
import { offerIdToHex } from '../../../utils/offer'

export const getTxDetailsTitle = (tx: TransactionSummary): string => {
  if (tx.offerId) {
    return i18n(tx.type === 'TRADE' ? 'wallet.stackedSats' : 'wallet.refund', offerIdToHex(tx.offerId))
  }
  return i18n(tx.type === 'WITHDRAWAL' ? 'wallet.withdrawal' : 'wallet.deposit')
}
