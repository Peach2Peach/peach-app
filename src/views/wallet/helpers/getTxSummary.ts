import { TransactionDetails } from 'bdk-rn/lib/classes/Bindings'
import { bitcoinStore } from '../../../store/bitcoinStore'
import { tradeSummaryStore } from '../../../store/tradeSummaryStore'
import { getTransactionType, txIsConfirmed } from '../../../utils/transaction'
import { walletStore } from '../../../utils/wallet/walletStore'

export const getTxSummary = (tx: TransactionDetails) => {
  const offerId = walletStore.getState().txOfferMap[tx.txid]
  const offer = tradeSummaryStore.getState().getOffer(offerId)
  const sats = Math.abs(tx.received - tx.sent)
  const price = sats / bitcoinStore.getState().satsPerUnit
  const type = getTransactionType(tx, offer)

  return {
    id: tx.txid,
    offerId,
    contractId: offer?.contractId,
    type,
    amount: sats,
    price,
    currency: bitcoinStore.getState().currency,
    date: txIsConfirmed(tx) ? new Date((tx.confirmationTime?.timestamp || Date.now()) * 1000) : new Date(),
    confirmed: txIsConfirmed(tx),
  }
}
