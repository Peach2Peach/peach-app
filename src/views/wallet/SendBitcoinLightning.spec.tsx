import { fireEvent, render, waitFor } from "test-utils";
import {
  lightningInvoice,
  lightningInvoiceNoAmount,
  lnPayment,
  lnPaymentFailed,
  nodeInfo,
} from "../../../tests/unit/data/lightningNetworkData";
import { queryClient } from "../../../tests/unit/helpers/QueryClientWrapper";
import { GlobalPopup } from "../../components/popup/GlobalPopup";
import { SendBitcoinLightning } from "./SendBitcoinLightning";

jest.mock("@breeztech/react-native-breez-sdk");
const sendPaymentMock = jest
  .requireMock("@breeztech/react-native-breez-sdk")
  .sendPayment.mockReturnValue({
    payment: lnPayment,
  });

jest.mock("@breeztech/react-native-breez-sdk");
const nodeInfoMock = jest
  .requireMock("@breeztech/react-native-breez-sdk")
  .nodeInfo.mockResolvedValue(nodeInfo);

jest.useFakeTimers();

describe("SendBitcoinLightning", () => {
  const { toJSON } = render(<SendBitcoinLightning />);
  const base = toJSON();

  afterEach(() => {
    queryClient.clear();
  });
  it("should render correctly", () => {
    expect(base).toMatchSnapshot();
  });

  it("should allow entering invoice", () => {
    const { getByAccessibilityHint } = render(<SendBitcoinLightning />);
    const $invoice = getByAccessibilityHint("lightning invoice");
    fireEvent.changeText($invoice, lightningInvoice);
    expect($invoice.props.value).toBe("lnbc1234 ... 4tl577");
  });
  it("should invalid invoice error for invalid invoices", () => {
    const { getByAccessibilityHint, getByText } = render(
      <SendBitcoinLightning />,
    );
    const $invoice = getByAccessibilityHint("lightning invoice");
    fireEvent.changeText($invoice, "invalid");
    expect(getByText("form.lightningInvoice.error")).toBeDefined();
  });
  it("should display amount input when invoice is set", () => {
    const { getByAccessibilityHint } = render(<SendBitcoinLightning />);
    const $invoice = getByAccessibilityHint("lightning invoice");
    fireEvent.changeText($invoice, lightningInvoice);
    const $amount = getByAccessibilityHint("amount");
    expect($amount).toBeDefined();
    expect($amount.props.value).toBe("12345");
  });
  it("should display amount and edit it  when invoice with no amount is set", () => {
    const amount = "20000";
    const { getByAccessibilityHint } = render(<SendBitcoinLightning />);
    const $invoice = getByAccessibilityHint("lightning invoice");
    fireEvent.changeText($invoice, lightningInvoiceNoAmount);
    const $amount = getByAccessibilityHint("amount");
    expect($amount).toBeDefined();
    fireEvent.changeText($amount, amount);
    expect($amount.props.value).toBe(amount);
  });
  it("should disable button when no invoice is set", () => {
    const { getByRole } = render(<SendBitcoinLightning />);
    const $payButton = getByRole("button");
    expect($payButton.props.accessibilityState.disabled).toBeTruthy();
  });
  it("should disable button not enough balance", () => {
    nodeInfoMock.mockResolvedValueOnce({
      ...nodeInfo,
      channelsBalanceMsat: 1000,
    });
    const { getByRole, getByAccessibilityHint } = render(
      <SendBitcoinLightning />,
    );
    const $payButton = getByRole("button");
    const $invoice = getByAccessibilityHint("lightning invoice");
    fireEvent.changeText($invoice, lightningInvoice);
    expect($payButton.props.accessibilityState.disabled).toBeTruthy();
  });

  it("should disable button when invoice amount is just 0", () => {
    const { getByRole, getByAccessibilityHint } = render(
      <SendBitcoinLightning />,
    );
    const $payButton = getByRole("button");
    const $invoice = getByAccessibilityHint("lightning invoice");
    fireEvent.changeText($invoice, lightningInvoiceNoAmount);
    expect($payButton.props.accessibilityState.disabled).toBeTruthy();
  });
  it("should disable button when user has enough balance", async () => {
    const { getByRole, getByAccessibilityHint } = render(
      <SendBitcoinLightning />,
    );
    const $payButton = getByRole("button");
    const $invoice = getByAccessibilityHint("lightning invoice");
    fireEvent.changeText($invoice, lightningInvoice);
    await waitFor(() =>
      expect($payButton.props.accessibilityState.disabled).toBeFalsy(),
    );
  });
  it("should pay invoice successfully", async () => {
    const { getByRole, getByAccessibilityHint } = render(
      <SendBitcoinLightning />,
    );
    const $invoice = getByAccessibilityHint("lightning invoice");
    const $payButton = getByRole("button");
    fireEvent.changeText($invoice, lightningInvoice);
    await waitFor(() =>
      expect($payButton.props.accessibilityState.disabled).toBeFalsy(),
    );
    fireEvent.press(getByRole("button"));
    const { getByText } = render(<GlobalPopup />);
    await waitFor(() => expect(getByText("invoice paid")).toBeDefined());
  });
  it("should display payment failure", async () => {
    sendPaymentMock.mockResolvedValueOnce({
      payment: lnPaymentFailed,
    });
    const { getByRole, getByAccessibilityHint } = render(
      <SendBitcoinLightning />,
    );
    const $invoice = getByAccessibilityHint("lightning invoice");
    const $payButton = getByRole("button");
    fireEvent.changeText($invoice, lightningInvoice);
    await waitFor(() =>
      expect($payButton.props.accessibilityState.disabled).toBeFalsy(),
    );
    fireEvent.press(getByRole("button"));
    const { getByText } = render(<GlobalPopup />);
    await waitFor(() => expect(getByText("failed to pay")).toBeDefined());
    expect(getByText("missing route")).toBeDefined();
  });
});
