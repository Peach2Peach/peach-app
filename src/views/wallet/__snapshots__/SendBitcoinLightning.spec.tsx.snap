// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SendBitcoinLightning should render correctly 1`] = `
<View
  onLayout={[Function]}
  style={
    [
      {
        "flexBasis": "0%",
        "flexGrow": 1,
        "flexShrink": 1,
      },
      {
        "paddingBottom": 0,
      },
    ]
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "flexBasis": "0%",
        "flexGrow": 1,
        "flexShrink": 1,
      }
    }
  >
    <View
      style={
        [
          {
            "borderBottomLeftRadius": 8,
            "borderBottomRightRadius": 8,
            "borderBottomWidth": 1,
          },
          {
            "paddingTop": 0,
            "zIndex": 1,
          },
          {
            "backgroundColor": "#FFF9F6",
          },
          {
            "borderColor": "#FEEDE5",
          },
        ]
      }
    >
      <View
        style={
          [
            {
              "alignItems": "center",
              "flexDirection": "row",
              "gap": 8,
              "paddingBottom": 6,
              "paddingLeft": 8,
              "paddingRight": 8,
              "paddingTop": 6,
            },
            {},
            undefined,
          ]
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "flexBasis": "0%",
              "flexDirection": "row",
              "flexGrow": 1,
              "flexShrink": 1,
              "gap": 4,
            }
          }
        >
          <Text
            allowFontScaling={false}
            numberOfLines={1}
            style={
              [
                {
                  "color": "#2B1911",
                  "fontFamily": "Baloo2-Regular",
                  "fontSize": 16,
                  "letterSpacing": 0.16,
                  "lineHeight": 24,
                },
                [
                  {
                    "fontFamily": "Baloo2-Bold",
                    "fontSize": 16,
                    "letterSpacing": 0.16,
                    "lineHeight": 26,
                  },
                  {},
                  {
                    "color": "#2B1911",
                  },
                  {
                    "flexBasis": "0%",
                    "flexGrow": 1,
                    "flexShrink": 1,
                  },
                ],
                false,
              ]
            }
          >
            send bitcoin
          </Text>
        </View>
        <View
          style={
            {
              "alignItems": "center",
              "flexDirection": "row",
              "gap": 10,
              "justifyContent": "flex-end",
            }
          }
        />
      </View>
    </View>
    <View
      style={
        [
          {
            "flexBasis": "0%",
            "flexGrow": 1,
            "flexShrink": 1,
          },
          false,
          {
            "paddingBottom": 0,
          },
        ]
      }
    >
      <View
        style={
          [
            {
              "flexBasis": "0%",
              "flexGrow": 1,
              "flexShrink": 1,
              "paddingBottom": 8,
              "paddingLeft": 8,
              "paddingRight": 8,
              "paddingTop": 8,
            },
            {},
            undefined,
          ]
        }
      >
        <View
          style={
            [
              {
                "gap": 8,
                "paddingBottom": 4,
                "paddingTop": 4,
              },
              {},
            ]
          }
        >
          <View>
            <Text
              allowFontScaling={false}
              style={
                [
                  {
                    "color": "#2B1911",
                    "fontFamily": "Baloo2-Regular",
                    "fontSize": 16,
                    "letterSpacing": 0.16,
                    "lineHeight": 24,
                  },
                  [
                    {
                      "fontFamily": "Baloo2-Regular",
                      "fontSize": 16,
                      "fontWeight": "bold",
                      "letterSpacing": 0.16,
                      "lineHeight": 23,
                      "paddingLeft": 8,
                    },
                    {
                      "color": "#2B1911",
                    },
                  ],
                  false,
                ]
              }
            >
              form.address.lightningInvoice.label
              <Text
                allowFontScaling={false}
                style={
                  [
                    {
                      "color": "#2B1911",
                      "fontFamily": "Baloo2-Regular",
                      "fontSize": 16,
                      "letterSpacing": 0.16,
                      "lineHeight": 24,
                    },
                    {
                      "color": "#BAADA7",
                    },
                    false,
                  ]
                }
              >
                 (optional)
              </Text>
            </Text>
            <View
              style={
                [
                  {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "gap": 4,
                    "justifyContent": "space-between",
                    "paddingLeft": 12,
                    "paddingRight": 12,
                    "width": "100%",
                  },
                  {
                    "borderRadius": 12,
                    "borderWidth": 1,
                    "overflow": "hidden",
                  },
                  {
                    "backgroundColor": "#FFFCFA",
                  },
                  {
                    "borderColor": "#2B1911",
                  },
                  false,
                  false,
                  {
                    "marginBottom": 1,
                    "marginTop": 1,
                  },
                  undefined,
                ]
              }
            >
              <TextInput
                accessibilityHint="form.address.lightningInvoice.label"
                allowFontScaling={false}
                autoCapitalize="none"
                autoCorrect={false}
                blurOnSubmit={false}
                editable={true}
                multiline={false}
                onBlur={[Function]}
                onChangeText={[Function]}
                onFocus={[Function]}
                placeholder="form.address.lightningInvoice.placeholder"
                placeholderTextColor="#EAE3DF"
                removeClippedSubviews={false}
                style={
                  [
                    {
                      "flexShrink": 1,
                      "fontFamily": "Baloo2-Regular",
                      "fontSize": 18,
                      "height": 40,
                      "letterSpacing": 0.18,
                      "lineHeight": 29.25,
                      "paddingBottom": 0,
                      "paddingTop": 0,
                      "width": "100%",
                    },
                    {
                      "color": "#BAADA7",
                    },
                    false,
                    {
                      "borderColor": "transparent",
                      "borderWidth": 1,
                    },
                    false,
                  ]
                }
                textAlignVertical="center"
                value=""
              />
              <View
                style={
                  {
                    "flexDirection": "row",
                    "gap": 16,
                  }
                }
              >
                <View
                  accessibilityState={
                    {
                      "busy": undefined,
                      "checked": undefined,
                      "disabled": undefined,
                      "expanded": undefined,
                      "selected": undefined,
                    }
                  }
                  accessibilityValue={
                    {
                      "max": undefined,
                      "min": undefined,
                      "now": undefined,
                      "text": undefined,
                    }
                  }
                  accessible={true}
                  collapsable={false}
                  focusable={true}
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    {
                      "opacity": 1,
                    }
                  }
                >
                  <test-file-stub
                    fill="#2B1911"
                    style={
                      [
                        {
                          "height": 20,
                          "width": 20,
                        },
                        undefined,
                      ]
                    }
                  />
                </View>
                <View
                  accessibilityState={
                    {
                      "busy": undefined,
                      "checked": undefined,
                      "disabled": undefined,
                      "expanded": undefined,
                      "selected": undefined,
                    }
                  }
                  accessibilityValue={
                    {
                      "max": undefined,
                      "min": undefined,
                      "now": undefined,
                      "text": undefined,
                    }
                  }
                  accessible={true}
                  collapsable={false}
                  focusable={true}
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    {
                      "opacity": 1,
                    }
                  }
                >
                  <test-file-stub
                    fill="#2B1911"
                    style={
                      [
                        {
                          "height": 20,
                          "width": 20,
                        },
                        undefined,
                      ]
                    }
                  />
                </View>
              </View>
            </View>
            <Text
              allowFontScaling={false}
              style={
                [
                  {
                    "color": "#2B1911",
                    "fontFamily": "Baloo2-Regular",
                    "fontSize": 16,
                    "letterSpacing": 0.16,
                    "lineHeight": 24,
                  },
                  [
                    {
                      "fontFamily": "Baloo2-Medium",
                      "fontSize": 14,
                      "letterSpacing": 0.14,
                      "lineHeight": 20,
                      "marginLeft": 12,
                      "marginTop": 4,
                    },
                    {
                      "color": "#DF321F",
                    },
                  ],
                  false,
                ]
              }
            >
               
            </Text>
          </View>
          <View
            accessibilityRole="button"
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": true,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignItems": "center",
                "backgroundColor": "#F56522",
                "borderColor": undefined,
                "borderRadius": 9999,
                "flexDirection": "row",
                "gap": 8,
                "justifyContent": "center",
                "minWidth": 104,
                "opacity": 0.33,
                "paddingBottom": 3,
                "paddingLeft": 24,
                "paddingRight": 24,
                "paddingTop": 3,
              }
            }
          >
            <Text
              allowFontScaling={false}
              ellipsizeMode="tail"
              numberOfLines={1}
              style={
                [
                  {
                    "color": "#2B1911",
                    "fontFamily": "Baloo2-Regular",
                    "fontSize": 16,
                    "letterSpacing": 0.16,
                    "lineHeight": 24,
                  },
                  [
                    {
                      "fontFamily": "Baloo2-SemiBold",
                      "fontSize": 13,
                      "letterSpacing": 0.47,
                      "lineHeight": 22,
                      "textTransform": "uppercase",
                    },
                    {},
                    {
                      "color": "#FFFCFA",
                    },
                  ],
                  false,
                ]
              }
            >
              wallet.sendBitcoin.payInvoice
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
